File: crates/rolldown/src/bundler/options/output_options.rs
use derivative::Derivative;

use super::file_name_template::FileNameTemplate;

#[derive(Debug)]
pub enum OutputFormat {
  Esm,
  Cjs,
}

#[derive(Debug)]
pub enum SourceMapType {
  File,
  Inline,
  Hidden,
}

impl From<String> for SourceMapType {
  fn from(value: String) -> Self {
    match value.as_str() {
      "file" => SourceMapType::File,
      "inline" => SourceMapType::Inline,
      "hidden" => SourceMapType::Hidden,
      _ => unreachable!("unknown sourcemap type"),
    }
  }
}

#[derive(Derivative)]
#[derivative(Debug)]
pub struct OutputOptions {
  pub entry_file_names: FileNameTemplate,
  pub chunk_file_names: FileNameTemplate,
  pub dir: String,
  pub format: OutputFormat,
  pub sourcemap: SourceMapType,
}

impl Default for OutputOptions {
  fn default() -> Self {
    Self {
      entry_file_names: FileNameTemplate::from("[name].js".to_string()),
      chunk_file_names: FileNameTemplate::from("[name]-[hash].js".to_string()),
      dir: "dist".into(),
      format: OutputFormat::Esm,
      sourcemap: SourceMapType::Hidden,
    }
  }
}

File: packages/rolldown/src/types/output-options.ts
import type { StringOrRegExp } from './utils'
import type { RenderedChunk, PreRenderedChunk } from '../binding'
import {
  SourcemapIgnoreListOption,
  SourcemapPathTransformOption,
} from '../rollup'

export type ModuleFormat =
  | 'es'
  | 'cjs'
  | 'esm'
  | 'module'
  | 'commonjs'
  | 'iife'
  | 'umd'

export type AddonFunction = (chunk: RenderedChunk) => string | Promise<string>

export type ChunkFileNamesFunction = (chunkInfo: PreRenderedChunk) => string

export interface OutputOptions {
  dir?: string
  file?: string
  exports?: 'auto' | 'named' | 'default' | 'none'
  format?: ModuleFormat
  sourcemap?: boolean | 'inline' | 'hidden'
  sourcemapIgnoreList?: boolean | SourcemapIgnoreListOption
  sourcemapPathTransform?: SourcemapPathTransformOption
  banner?: string | AddonFunction
  footer?: string | AddonFunction
  intro?: string | AddonFunction
  outro?: string | AddonFunction
  extend?: boolean
  esModule?: boolean | 'if-default-prop'
  assetFileNames?: string
  entryFileNames?: string | ChunkFileNamesFunction
  chunkFileNames?: string | ChunkFileNamesFunction
  cssEntryFileNames?: string | ChunkFileNamesFunction
  cssChunkFileNames?: string | ChunkFileNamesFunction
  minify?: boolean
  name?: string
  globals?: Record<string, string>
  externalLiveBindings?: boolean
  inlineDynamicImports?: boolean
  advancedChunks?: {
    minSize?: number
    minShareCount?: number
    groups?: {
      name: string
      test?: StringOrRegExp
      priority?: number
      minSize?: number
      minShareCount?: number
    }[]
  }
  /**
   * Control comments in the output.
   *
   * - `none`: no comments
   * - `preserve-legal`: preserve comments that contain `@license`, `@preserve` or starts with `//!` `/*!`
   */
  comments?: 'none' | 'preserve-legal'
}

interface OverwriteOutputOptionsForCli {
  banner?: string
  footer?: string
  intro?: string
  outro?: string
  esModule?: boolean
  advancedChunks?: {
    minSize?: number
    minShareCount?: number
  }
}

export type OutputCliOptions = Omit<
  OutputOptions,
  | keyof OverwriteOutputOptionsForCli
  | 'sourcemapIgnoreList'
  | 'sourcemapPathTransform'
> &
  OverwriteOutputOptionsForCli
